src/source/Ice/SocketConnection.c

```diff
@@ -155,37 +160,75 @@ STATUS socketConnectionSendData(PSocketConnection pSocketConnection, PBYTE pBuf,
     struct sockaddr_in6 ipv6Addr;
     fd_set wfds;
     struct timeval tv;
+    SIZE_T wBioDataLen = 0;
+    PCHAR wBioBuffer = NULL;
 
-    CHK(pSocketConnection != NULL && pBuf != NULL, STATUS_NULL_ARG);
-    CHK(bufLen != 0 && (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP || pDestIp != NULL), STATUS_INVALID_ARG);
+    CHK(pSocketConnection != NULL, STATUS_NULL_ARG);
+    CHK((pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP || pDestIp != NULL), STATUS_INVALID_ARG);
     CHK_WARN(!ATOMIC_LOAD_BOOL(&pSocketConnection->connectionClosed), STATUS_SOCKET_CONNECTION_CLOSED_ALREADY, "Failed to send data. Socket closed already");
 
     MUTEX_LOCK(pSocketConnection->lock);
     locked = TRUE;
 
     if (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP && pSocketConnection->secureConnection) {
-        // underlying BIO has been bound to socket so SSL_write sends the encrypted data.
-        // because we are using non-blocking socket, SSL_write may return SSL_ERROR_WANT_WRITE in which case
-        // SSL_write should be called again with the same buffer.
-        while(iterate) {
+        if (SSL_is_init_finished(pSocketConnection->pSsl)) {
+            /* Should have a valid buffer */
+            CHK(pBuf != NULL && bufLen > 0, STATUS_INVALID_ARG);
             sslRet = SSL_write(pSocketConnection->pSsl, pBuf, bufLen);
-            if (sslRet > 0 || (sslErr = SSL_get_error(pSocketConnection->pSsl, sslRet)) != SSL_ERROR_WANT_WRITE) {
-                iterate = FALSE;
-            } else {
-                THREAD_SLEEP(SSL_WRITE_RETRY_DELAY);
+            if (sslRet < 0){
+                sslErr = SSL_get_error(pSocketConnection->pSsl, sslRet);
+                switch (sslErr) {
+                    case SSL_ERROR_WANT_READ:
+                        /* explicit fall-through */
+                    case SSL_ERROR_WANT_WRITE:
+                        break;
+                    default:
+                        DLOGW("SSL_write failed with %s", ERR_error_string(sslErr, NULL));
+                        DLOGD("Close socket %d", pSocketConnection->localSocket);
+                        ATOMIC_STORE_BOOL(&pSocketConnection->connectionClosed, TRUE);
+                        break;
+                }
+
+                CHK(FALSE, STATUS_SEND_DATA_FAILED);
```

Comment:
Do we want to error out with STATUS_SEND_DATA_FAILED on want read/write?

Reply:
the STATUS_SEND_DATA_FAILED does get handled by its callers.